### DISCUSSION TITLE
Inline Test Support

### DISCUSSION THREAD

**Discussion by calebmeyer** [zed-industries/zed#6804] [Feature Requests]
One thing that I miss coming from VS Code is the ability to run tests (and see whether they're passing or failing) inline. This is done with a gutter icon next to every `describe`, `it`, or `test` (in jest, for JS).
<img width="263" alt="image" src="https://github.com/zed-industries/zed/assets/3342777/49174edf-7795-43ee-8d28-6acb66684670">

When I save a change to a test file, only that file is re-run, and failing tests are shown inline:
<img width="1303" alt="image" src="https://github.com/zed-industries/zed/assets/3342777/47928902-924b-4445-a074-51c4bf2845a9">

This is a really smooth workflow for writing tests, since you get immediate feedback about both passing and failing tests.

I'd love to see something like this implemented in a generic way (not just for `jest`, but also `cargo test`, `go test`, `rspec` etc) for Zed. I know that running things without the terminal is probably a pre-requisite, but now that vim support is good enough, this is my last remaining blocker to switching from VS Code.

**Comment by d1y** [comment]
I need this too much

**Comment by ScopeSV** [comment]
agreed. Running tests within the editor, and also good git support and I would say bye bye to vscode.

**Comment by dannyfuf** [comment]
If I could run tests with shortcuts, I would also leave VSCode

**Comment by konstantinzolotarev** [comment]
Same here.
The only feature I wait to leave VSCode

**Comment by Moshyfawn** [comment]
#7108

**Reply by dandv** [reply]
Tasks have been implemented. I've read the docs, but [haven't been able to get any task to run](https://github.com/zed-industries/zed/issues/5242#issuecomment-2523199757).

**Comment by TomVance** [comment]
Not quite as nice as inline tests (I'd love this too). But I wrote up some quick shortcuts using Zed tasks (https://zed.dev/docs/tasks) to make these a bit less painful for me for use with Rspec, but pretty easy to swap to other tooling

```json
[
  {
    "label": "Run Rspec",
    "command": "echo $ZED_FILE | sed 's|app/|spec/|g' | sed 's|.rb|_spec.rb|g' | sed 's/_spec_spec.rb/_spec.rb/g' | { read spec; bundle exec rspec $spec }",
    "use_new_terminal": false
  },
  {
    "label": "Jump to Spec",
    "command": "echo $ZED_FILE | sed 's|app/|spec/|g' | sed 's|.rb|_spec.rb|g' | { read spec; zed $spec }",
    "reveal": "never",
    "use_new_terminal": false
  }
]
```
  
Please don't judge some of my poor bashing here, was just a quick thing to get moving with Zed

**Comment by dannyfuf** [comment]
Since Zed Tasks allows us to run any command, I wrote a quick Python script to run specific jest tests based on the cursor position with shortcuts. If interested checkout [here](https://github.com/dannyfuf/zed-test-runner/tree/main). I think it is easy to extend to support other languages.

**Comment by Moshyfawn** [comment]
JS/TS Test Runbbles support has landed in [Zed Preview `0.139.3`](https://zed.dev/releases/preview/0.139.3) 👀

**Reply by samueljim** [reply]
is there support for changing the test command? I would like to run a setup file

**Reply by AchilleBailly** [reply]
Hey, I've had a look around in the code and there is a way to do it !

For that you have to find the name of the tag that is used for the inline launch. You can find them in the `runnables.scm` file for each new language support. For natively supported languages, they are in the `crates/languages/src/<language>/` folder. Otherwise, if it is an extension, you have to look for `extensions/<language>/languages/<language>/`. 

For example, Javascript uses `ts-test` tag and you can overriade the default command with this added in the `tasks.json` file:
```
{
    "label": "JavaScript Test",
    "command": "<new command to launch>",
    "args": ["\"$ZED_RELATIVE_FILE\""],
    "tags": ["js-test"]
  },
  ```

**Reply by onx2** [reply]
**EDIT** - I fixed my error.


This is great thank you! 

```
{
  "label": "Run Vitest",
  "command": "bun run vitest run",
  "args": ["\"$ZED_RELATIVE_FILE\" -t=\"$ZED_SYMBOL\""],
  "tags": ["tsx-test", "ts-test"]
}
```

[Vitest docs](https://vitest.dev/guide/cli#testnamepattern)
[Zed docs](https://zed.dev/docs/tasks#variables)

**Reply by AlbertMarashi** [reply]
Awesome thank you for this @onx2 - works perfectly with the zed "Play" test button

![image](https://github.com/user-attachments/assets/2e0f5bd7-652d-4772-9a5d-dad6e2dd42e0)

**Reply by dandv** [reply]
I tried to run Deno tests but nothing happens when I click the play button. [What am I doing wrong?](https://github.com/zed-industries/zed/issues/5242#issuecomment-2523199757)

**Comment by avramukk** [comment]
@AlbertMarashi Is it possible to do the same running icons for my playwright testing?
![CleanShot 2024-11-30 at 00 34 11@2x](https://github.com/user-attachments/assets/15446950-2fd4-49e7-809f-f7a2d3dab3ad)

**Reply by AlbertMarashi** [reply]
Yes, you will need to do something similar to https://github.com/zed-industries/zed/discussions/6804#discussioncomment-9912007

**Comment by Jimmy-Z** [comment]
This is the thing I miss the most from vscode.

**Comment by somtooo** [comment]
Any update on weather we can see tests passing or failing inline?

**Comment by Ruegen** [comment]
I would like this feature too

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/community-links

**Reply by dandv** [reply]
How is the community better supported, when the information shared through Discord is [inaccessible outside](https://www.reddit.com/r/LocalLLaMA/comments/1ao4aio/comment/kpwr64r/)? Does the team enjoy answering the same questions over and over?

Back on topic, it seems inline test support hasn't been implemented (at least not for Deno) - #5242.
